examples/all-devices-app/esp32/main/main.cpp

```diff
@@ -0,0 +1,369 @@
+    }
+
+    err = Server::GetInstance().Init(initParams);
+    if (err != CHIP_NO_ERROR)
+    {
+        ESP_LOGE(TAG, "Server init failed: %" CHIP_ERROR_FORMAT, err.Format());
+        return;
+    }
+
+#if CHIP_DEVICE_CONFIG_ENABLE_WIFI && CONFIG_ENABLE_CHIP_SHELL
+    chip::Shell::SetWiFiDriver(&sWiFiDriver);
+#endif
+}
+
+} // namespace
+
+void InitServerWithDeviceType(std::string deviceType)
+{
+    // Set the device type (store the actual string, not a pointer to temporary)
+    gDeviceType = std::move(deviceType);
+
+    // Init the server
+    PlatformMgr().ScheduleWork(InitServer, reinterpret_cast<intptr_t>(nullptr));
+}
+
+extern "C" void app_main()
+{
+    // Initialize the ESP NVS layer.
+    esp_err_t err = nvs_flash_init();
+    if (err != ESP_OK)
+    {
+        ESP_LOGE(TAG, "nvs_flash_init() failed: %s", esp_err_to_name(err));
+        return;
+    }
+    err = esp_event_loop_create_default();
+    if (err != ESP_OK)
+    {
+        ESP_LOGE(TAG, "esp_event_loop_create_default() failed: %s", esp_err_to_name(err));
+        return;
+    }
+
+    ESP_LOGI(TAG, "==================================================");
+    ESP_LOGI(TAG, "chip-esp32-all-devices-app starting");
+    ESP_LOGI(TAG, "==================================================");
+
+#if CHIP_DEVICE_CONFIG_ENABLE_WIFI
+    if (DeviceLayer::Internal::ESP32Utils::InitWiFiStack() != CHIP_NO_ERROR)
+    {
+        ESP_LOGE(TAG, "Failed to initialize WiFi stack");
+        return;
+    }
+#endif // CHIP_DEVICE_CONFIG_ENABLE_WIFI
+
+    DeviceLayer::SetDeviceInfoProvider(&gExampleDeviceInfoProvider);
+
+    CHIP_ERROR error = chip::Platform::MemoryInit();
+    if (error != CHIP_NO_ERROR)
+    {
+        ESP_LOGE(TAG, "Platform::MemoryInit() failed: %" CHIP_ERROR_FORMAT, error.Format());
+        return;
+    }
+
+    error = PlatformMgr().InitChipStack();
+    if (error != CHIP_NO_ERROR)
+    {
+        ESP_LOGE(TAG, "PlatformMgr().InitChipStack() failed: %" CHIP_ERROR_FORMAT, error.Format());
+        return;
+    }
+
+    // Register event handler to restart DNS-SD on connectivity change
+    PlatformMgr().AddEventHandler(DeviceEventHandler, 0);
+
+#if CONFIG_ENABLE_ESP32_FACTORY_DATA_PROVIDER
+    SetCommissionableDataProvider(&sFactoryDataProvider);
+#if CONFIG_ENABLE_ESP32_DEVICE_INSTANCE_INFO_PROVIDER
+    SetDeviceInstanceInfoProvider(&sFactoryDataProvider);
+#endif
+    SetDeviceAttestationCredentialsProvider(&sFactoryDataProvider);
+#else
+    SetDeviceAttestationCredentialsProvider(Examples::GetExampleDACProvider());
+#endif // CONFIG_ENABLE_ESP32_FACTORY_DATA_PROVIDER
+
+    error = PlatformMgr().StartEventLoopTask();
+    if (error != CHIP_NO_ERROR)
+    {
+        ESP_LOGE(TAG, "PlatformMgr().StartEventLoopTask() failed: %" CHIP_ERROR_FORMAT, error.Format());
+        return;
+    }
+
+#if CONFIG_ENABLE_CHIP_SHELL
+    chip::LaunchShell();
+    chip::Shell::DeviceCommands::GetInstance().Register();
```

Comment:
Can this be moved into the below `else...` condition? I mean only register these if we are do not have a device type stored